### PR TITLE
Add support for AVRO in Iceberg

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Add support for AVRO in Iceberg
Supersedes #4776

## Documentation

(x) Sufficient documentation is included in this PR.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Add support for Avro file format. ({issue}`12125`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java [hunk 1]
```diff
@@ -849,6 +868,120 @@ else if (column.isRowPositionColumn()) {
...
+                    columnNames.add(column.getName());
+                    columnTypes.add(column.getType());
+
+                    if (field == null) {
+                        constantPopulatingPageSourceBuilder.addConstantColumn(nativeValueToBlock(column.getType(), null));
+                    }
+                    else {
+                        constantPopulatingPageSourceBuilder.addDelegateColumn(avroSourceChannel);
+                    }
+                    avroSourceChannel++;
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java:931]: I think these need to be in the else block?
```suggestion
                    if (field == null) {
                        constantPopulatingPageSourceBuilder.addConstantColumn(nativeValueToBlock(column.getType(), null));
                    }
                    else {
                        rowIndexChannels.add(false);
                        columnNames.add(column.getName());
                        columnTypes.add(column.getType());
                        constantPopulatingPageSourceBuilder.addDelegateColumn(avroSourceChannel);
                        avroSourceChannel++;
                    }
```

**Review by alexjo2144 — COMMENTED** [review]
Just an initial skim. Looks like Avro table stat's aren't working. Is that because Avro doesn't have them, or because the read path just isn't implemented for them yet?

**Comment by ebyhr** [comment]
> Is that because Avro doesn't have them, or because the read path just isn't implemented for them yet?

Avro file/table doesn't support statistics as far as I know.

**Review by electrum — APPROVED** [review]
Apologies for the long review time. Let's get this in since it's needed for https://github.com/trinodb/trino/pull/13219

**Comment by ebyhr** [comment]
Just rebased on upstream to resolve a logical conflict by 308e717baf78dc344d7f962bf32155416b751902

**Comment by electrum** [comment]
Looks like a real test failure:
```
2022-07-22 06:58:45 INFO: FAILURE     /    io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.testSparkReadingTrinoCompressedData [AVRO, ZSTD] (Groups: profile_specific_tests, iceberg) took 0.6 seconds
2022-07-22 06:58:45 SEVERE: Failure cause:
java.lang.AssertionError: Expected callback to throw QueryExecutionException
	at io.trino.tempto.assertions.QueryAssert.assertQueryFailure(QueryAssert.java:107)
	at io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.testSparkReadingTrinoCompressedData(TestIcebergSparkCompatibility.java:1238)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

**Pull request merged**